File: backend/src/acc_nomad/services/pmg_delivery.py
"""Envio de relatório PMG por Email e WhatsApp."""

from __future__ import annotations

import smtplib
from datetime import date, datetime, timezone
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Any

from acc_nomad.config import settings
from acc_nomad.services.supabase_client import get_supabase_client
from acc_nomad.services.whatsapp_client import WhatsAppError, send_whatsapp_message


class DeliveryError(RuntimeError):
    pass


def _format_pmg_text(resumo: dict[str, float], periodo: str, empresa: str) -> str:
    resultado = (
        resumo.get("receita", 0)
        - resumo.get("custo_variavel", 0)
        - resumo.get("custo_fixo", 0)
        - resumo.get("juros", 0)
        - resumo.get("investimentos", 0)
    )
    return f"""ACC Nomad — Relatório PMG
Empresa: {empresa}
Período: {periodo}

Receita:          R$ {resumo.get('receita', 0):,.2f}
Custo Variável:   R$ {resumo.get('custo_variavel', 0):,.2f}
Custo Fixo:       R$ {resumo.get('custo_fixo', 0):,.2f}
Juros:            R$ {resumo.get('juros', 0):,.2f}
Investimentos:    R$ {resumo.get('investimentos', 0):,.2f}
─────────────────────────
Resultado:        R$ {resultado:,.2f}

Gerado automaticamente pelo ACC Nomad.
"""


def send_email(destinatario: str, assunto: str, corpo: str) -> None:
    if not settings.smtp_host or not settings.email_from:
        raise DeliveryError(
            "SMTP não configurado. Defina SMTP_HOST, SMTP_USER, SMTP_PASSWORD e EMAIL_FROM em backend/.env"
        )

    msg = MIMEMultipart()
    msg["From"] = settings.email_from
    msg["To"] = destinatario
    msg["Subject"] = assunto
    msg.attach(MIMEText(corpo, "plain", "utf-8"))

    with smtplib.SMTP(settings.smtp_host, settings.smtp_port) as server:
        if settings.smtp_use_tls:
            server.starttls()
        if settings.smtp_user and settings.smtp_password:
            server.login(settings.smtp_user, settings.smtp_password)
        server.send_message(msg)


def registrar_envio(
    *,
    empresa_id: str,
    periodo_inicio: date,
    periodo_fim: date,
    canal: str,
    destinatario: str,
    status: str,
    enviado_por: str | None = None,
    erro_mensagem: str | None = None,
    provider_name: str | None = None,
    provider_message_id: str | None = None,
    confirmado_em: str | None = None,
) -> dict[str, Any]:
    client = get_supabase_client()
    row = {
        "empresa_id": empresa_id,
        "periodo_inicio": periodo_inicio.isoformat(),
        "periodo_fim": periodo_fim.isoformat(),
        "canal": canal,
        "destinatario": destinatario,
        "status": status,
        "enviado_por": enviado_por,
        "erro_mensagem": erro_mensagem,
        "provider_name": provider_name,
        "provider_message_id": provider_message_id,
        "confirmado_em": confirmado_em,
    }
    result = client.table("envios_pmg").insert(row).execute()
    return {"saved": True, "data": result.data}


async def enviar_pmg(
    *,
    empresa_id: str,
    empresa_nome: str,
    periodo: str,
    periodo_inicio: date,
    periodo_fim: date,
    canal: str,
    destinatario: str,
    resumo: dict[str, float],
    enviado_por: str | None = None,
) -> dict[str, str]:
    corpo = _format_pmg_text(resumo, periodo, empresa_nome)
    assunto = f"ACC Nomad — PMG {periodo} — {empresa_nome}"

    try:
        provider_name: str | None = None
        provider_message_id: str | None = None
        destinatario_final = destinatario

        if canal == "email":
            send_email(destinatario, assunto, corpo)
        elif canal == "whatsapp":
            try:
                wa = send_whatsapp_message(destinatario, corpo)
            except WhatsAppError as exc:
                raise DeliveryError(str(exc)) from exc
            provider_name = wa.get("provider")
            provider_message_id = wa.get("message_id")
            destinatario_final = wa.get("normalized_number", destinatario)
        else:
            raise DeliveryError(f"Canal inválido: {canal}")

        confirmado_em = datetime.now(timezone.utc).isoformat()

        registrar_envio(
            empresa_id=empresa_id,
            periodo_inicio=periodo_inicio,
            periodo_fim=periodo_fim,
            canal=canal,
            destinatario=destinatario_final,
            status="enviado",
            enviado_por=enviado_por,
            provider_name=provider_name,
            provider_message_id=provider_message_id,
            confirmado_em=confirmado_em,
        )
        return {
            "status": "enviado",
            "canal": canal,
            "destinatario": destinatario_final,
            "provider_name": provider_name,
            "provider_message_id": provider_message_id,
        }
    except Exception as exc:
        registrar_envio(
            empresa_id=empresa_id,
            periodo_inicio=periodo_inicio,
            periodo_fim=periodo_fim,
            canal=canal,
            destinatario=destinatario,
            status="erro",
            enviado_por=enviado_por,
            erro_mensagem=str(exc),
        )
        raise DeliveryError(str(exc)) from exc

File: frontend/src/lib/server/pmg-delivery.ts
import nodemailer from "nodemailer";
import type { PmgResumo } from "@/lib/types";
import { getSupabaseAdmin } from "@/lib/server/supabase-admin";

export class PmgDeliveryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PmgDeliveryError";
  }
}

function formatPmgText(resumo: PmgResumo, periodo: string, empresa: string): string {
  const resultado =
    resumo.receita -
    resumo.custo_variavel -
    resumo.custo_fixo -
    resumo.juros -
    resumo.investimentos;

  return `ACC Nomad — Relatório PMG
Empresa: ${empresa}
Período: ${periodo}

Receita:          R$ ${resumo.receita.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}
Custo Variável:   R$ ${resumo.custo_variavel.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}
Custo Fixo:       R$ ${resumo.custo_fixo.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}
Juros:            R$ ${resumo.juros.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}
Investimentos:    R$ ${resumo.investimentos.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}
─────────────────────────
Resultado:        R$ ${resultado.toLocaleString("pt-BR", { minimumFractionDigits: 2 })}

Gerado automaticamente pelo ACC Nomad.
`;
}

async function sendEmail(destinatario: string, assunto: string, corpo: string): Promise<void> {
  const host = process.env.SMTP_HOST;
  const from = process.env.EMAIL_FROM;

  if (!host || !from) {
    throw new PmgDeliveryError(
      "SMTP não configurado na Vercel. Defina SMTP_HOST, SMTP_USER, SMTP_PASSWORD e EMAIL_FROM.",
    );
  }

  const port = Number(process.env.SMTP_PORT || 587);
  const transporter = nodemailer.createTransport({
    host,
    port,
    secure: port === 465,
    auth: process.env.SMTP_USER
      ? {
          user: process.env.SMTP_USER,
          pass: process.env.SMTP_PASSWORD || "",
        }
      : undefined,
  });

  await transporter.sendMail({
    from,
    to: destinatario,
    subject: assunto,
    text: corpo,
  });
}

async function sendWhatsApp(destinatario: string, mensagem: string): Promise<void> {
  const url = process.env.WHATSAPP_API_URL;
  if (!url) {
    throw new PmgDeliveryError(
      "WhatsApp não configurado. Defina WHATSAPP_API_URL e WHATSAPP_API_TOKEN na Vercel.",
    );
  }

  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (process.env.WHATSAPP_API_TOKEN) {
    headers.Authorization = `Bearer ${process.env.WHATSAPP_API_TOKEN}`;
  }

  const res = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify({ number: destinatario, message: mensagem }),
  });

  if (!res.ok) {
    const text = (await res.text()).slice(0, 200);
    throw new PmgDeliveryError(`WhatsApp API erro ${res.status}: ${text}`);
  }
}

async function registrarEnvio(params: {
  empresa_id: string;
  periodo_inicio: string;
  periodo_fim: string;
  canal: string;
  destinatario: string;
  status: string;
  enviado_por?: string | null;
  erro_mensagem?: string | null;
}) {
  const supabase = getSupabaseAdmin();
  await supabase.from("envios_pmg").insert({
    empresa_id: params.empresa_id,
    periodo_inicio: params.periodo_inicio,
    periodo_fim: params.periodo_fim,
    canal: params.canal,
    destinatario: params.destinatario,
    status: params.status,
    enviado_por: params.enviado_por ?? null,
    erro_mensagem: params.erro_mensagem ?? null,
  });
}

export async function enviarPmg(params: {
  empresa_id: string;
  empresa_nome: string;
  periodo: string;
  periodo_inicio: string;
  periodo_fim: string;
  canal: "email" | "whatsapp";
  destinatario: string;
  resumo: PmgResumo;
  enviado_por?: string | null;
}) {
  const corpo = formatPmgText(params.resumo, params.periodo, params.empresa_nome);
  const assunto = `ACC Nomad — PMG ${params.periodo} — ${params.empresa_nome}`;

  try {
    if (params.canal === "email") {
      await sendEmail(params.destinatario, assunto, corpo);
    } else if (params.canal === "whatsapp") {
      await sendWhatsApp(params.destinatario, corpo);
    } else {
      throw new PmgDeliveryError(`Canal inválido: ${params.canal}`);
    }

    await registrarEnvio({
      empresa_id: params.empresa_id,
      periodo_inicio: params.periodo_inicio,
      periodo_fim: params.periodo_fim,
      canal: params.canal,
      destinatario: params.destinatario,
      status: "enviado",
      enviado_por: params.enviado_por,
    });

    return { status: "enviado", canal: params.canal, destinatario: params.destinatario };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    await registrarEnvio({
      empresa_id: params.empresa_id,
      periodo_inicio: params.periodo_inicio,
      periodo_fim: params.periodo_fim,
      canal: params.canal,
      destinatario: params.destinatario,
      status: "erro",
      enviado_por: params.enviado_por,
      erro_mensagem: message,
    });
    throw new PmgDeliveryError(message);
  }
}
